### DISCUSSION TITLE
使用 WASM 在浏览器中合并视频和音频

### DISCUSSION THREAD

**Discussion by WakelessSloth56** [the1812/Bilibili-Evolved#4518] [功能建议 / Ideas]
使用 WASM 在浏览器中合并 DASH 的视频和音频。
最近在写别的项目的 WASM 时突然有这么个想法，应该可以写一个视频下载的插件集成到 [Bilibili-Evolved](https://github.com/the1812/Bilibili-Evolved) 中。
**我有时间的话会尝试做一做，先开个讨论提供一下这个思路。**

---

目前我想到的问题：
- WASM 最大内存 2GB，太大的文件搞不了
- 需要先在浏览器中下载音/视频轨然后合并，期间不能关闭标签页
- ~~WASM 文件考虑到网络环境是否需要缓存？~~

**Comment by the1812** [comment]
> WASM 文件考虑到网络环境是否需要缓存？

缓存什么内容?

**Reply by WakelessSloth56** [reply]
.wasm文件

**Reply by the1812** [reply]
目前动态载入的库都是实时下载的

**Reply by WakelessSloth56** [reply]
OK

**Comment by WakelessSloth56** [comment]
一上来就遇到问题了，没办法跨域加载 Web Worker：
而且完全搞不懂为什么会是 `https://s1.hdslb.com/bfs/seed/0.js` ……
```
index.ts:142 Uncaught (in promise) DOMException: Failed to construct 'Worker': Script at 'https://s1.hdslb.com/bfs/seed/0.js' cannot be accessed from origin 'https://www.bilibili.com'.
    at FFmpeg.eval [as load] (webpack://@bevo/core/./registry/lib/plugins/video/download/wasm-output/ffmpeg/index.ts?:142:47)
    at load (webpack://@bevo/core/./registry/lib/plugins/video/download/wasm-output/handler.ts?:26:16)
    at async trans (webpack://@bevo/core/./registry/lib/plugins/video/download/wasm-output/handler.ts?:34:5)
```

```typescript
new Worker(new URL('./worker.ts', import.meta.url), {
    type: 'module',
})
```
就变成了这样：
```javascript
new Worker(new URL(/* worker import */ __webpack_require__.p + __webpack_require__.u(0), __webpack_require__.b)
```
`__webpack_require__.p` 值是 `https://s1.hdslb.com/bfs/seed/jinkela/commentpc/../../`

大概应该是 Webpack 的问题：
- https://github.com/webpack/webpack/discussions/14648
- https://github.com/webpack/webpack/issues/16696

也可能是我没摸清 Bilibili-Evolved 的打包配置，我平时是不怎么用 Webpack 的emmmmmm

**Reply by the1812** [reply]
我也没用过 webpack + Web Worker, 打包配置感觉跟这个有关系的地方就是限制了不能分 chunk (不能拆分多文件, 否则油猴加载不了)
看了下 issue 可能的一个解决方式就是把 worker 代码加载到一个 Blob 里, new Worker 的时候用 Blob URL 载入

**Comment by WakelessSloth56** [comment]
![‎2023‎11‎13‎‏223935](https://github.com/the1812/Bilibili-Evolved/assets/43951496/a187fa4f-2e85-4a2a-b5a4-3b46bd2f0696)

---

功能已经实现了 ~~，但是实现方法比较奇怪~~：
上面提到的 Blob 方法不可行，因为 `ffmpeg.wasm` 这个项目的 Worker 的特殊，目前想加载外部的 Worker 得替换掉一个地方的代码（参见：https://github.com/ffmpegwasm/ffmpeg.wasm/issues/548 ，https://github.com/ffmpegwasm/ffmpeg.wasm/discussions/580 ，https://github.com/ffmpegwasm/ffmpeg.wasm/pull/562 ）， ~~所以目前我的方案是类似于加载 `lodash` 那样作为全局加载：从 Tampermonkey 里 `@require` 一个修改过的 JS 文件。~~
~~这种做法并不好，只能算是临时方案，可以等待一下上游的后续更新。~~
~~但是忽略这一点的话，对功能的实现已经可以了。~~

`ffmpeg.wasm` 是使用的 MIT 许可证，我修改了它来实现所需的功能，并对用不到的功能进行了精简。
我会在整理优化后发起 PR。

**Reply by WakelessSloth56** [reply]
最初我是想我来写一个 WASM 的，对于只混流而已用 `ffmpeg.wasm` 太大了，后来因为个人时间的原因还是先用了它。

**Reply by the1812** [reply]
那应该使用脚本提供的 `RuntimeLibrary` 功能来加载, `@require` 只是用来加载本体必须的依赖

**Reply by WakelessSloth56** [reply]
已经解决了，那个只是个尝试。
